nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,455 @@
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	clusterConfig, err := client.Get()
+	if err != nil {
+		return fmt.Errorf("Failed to read Nsxt Manager Cluster: %s", err)
+	}
+	nodeInfo := clusterConfig.Nodes
+	var nodes []NsxClusterNode
+	hostIP, err := getHostIPFromClusterConfig(m, clusterConfig)
+	isIPv4 := (net.ParseIP(hostIP)).To4() != nil
+	if err != nil {
+		return err
+	}
+	for _, node := range nodeInfo {
+		ip := getIPFromNodeInfo(node, isIPv4)
+		fqdn := *node.Fqdn
+		status := *node.Status
+		if ip != hostIP {
+			id := *node.NodeUuid
+			clusterNode := NsxClusterNode{
+				ID:        id,
+				IPAddress: ip,
+				Fqdn:      fqdn,
+				Status:    status,
+			}
+			nodes = append(nodes, clusterNode)
+		}
+
+	}
+	d.Set("revision", clusterConfig.Revision)
+	setNodesInSchema(d, nodes)
+	return nil
+}
+
+func getIPFromNodeInfo(node nsxModel.ClusterNodeInfo, isIPv4 bool) string {
+	// After join node into cluster, apiListen address may take some time to become available
+	// if it's not available, we retrieve node ip from nodeEntityInfo
+	var ip *string
+	apiListenAddr := node.ApiListenAddr
+	if apiListenAddr != nil {
+		if isIPv4 {
+			ip = apiListenAddr.IpAddress
+		} else {
+			ip = apiListenAddr.Ipv6Address
+		}
+	} else {
+		nodeEntities := node.Entities
+		for _, entity := range nodeEntities {
+			if *entity.Port == 443 {
+				if isIPv4 {
+					ip = entity.IpAddress
+				} else {
+					ip = entity.Ipv6Address
+				}
+				break
+			}
+		}
+	}
+	return *ip
+}
+
+func resourceNsxtManagerClusterUpdate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+
+	clusterID, certSha256Thumbprint, hostIP, err := getClusterInfoFromHostNode(d, m)
+	if err != nil {
+		return err
+	}
+	oldNodes, newNodes := d.GetChange("node")
+	oldNodesIPs := getClusterNodesIPs(oldNodes)
+	newNodesIPs := getClusterNodesIPs(newNodes)
+	for _, node := range oldNodes.([]interface{}) {
+		nodeMap := node.(map[string]interface{})
+		ip := nodeMap["ip_address"].(string)
+		if !slices.Contains(newNodesIPs, ip) {
+			id := nodeMap["id"].(string)
+			force := "true"
+			gracefulShutdown := "true"
+			ignoreRepositoryIPCheckParam := "false"
+			_, err := client.Removenode(id, &force, &gracefulShutdown, &ignoreRepositoryIPCheckParam)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for _, node := range newNodes.([]interface{}) {
+		nodeMap := node.(map[string]interface{})
+		ip := nodeMap["ip_address"].(string)
+		if !slices.Contains(oldNodesIPs, ip) {
+			userName := nodeMap["username"].(string)
+			password := nodeMap["password"].(string)
+			ip := nodeMap["ip_address"].(string)
+			nodeObj := NsxClusterNode{
+				IPAddress: ip,
+				UserName:  userName,
+				Password:  password,
+			}
+			err = joinNodeToCluster(clusterID, certSha256Thumbprint, nodeObj, hostIP, d, m)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
```

Comment:
I wonder what should happen if username or password are updated on a node? I think according to this code no API calls would happen, which is probably correct, since we don't need to rejoin.

Reply:
I think nothing will happen in that case, if they are already in cluster then any change in username/password will not affect them. If we want to delete them, it still does not matter because the remove node api does not require guest node credential